realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java

```diff
@@ -106,4 +109,69 @@ public void uploadDownloadAllChanges() throws InterruptedException {
         assertEquals(1, adminRealm.where(AllTypes.class).count());
         adminRealm.close();
     }
+
+    @Test
+    public void interruptWaits() throws InterruptedException {
+        final SyncUser user = UserFactory.createUniqueUser(Constants.AUTH_URL);
+        SyncUser adminUser = UserFactory.createAdminUser(Constants.AUTH_URL);
+        final SyncConfiguration userConfig = configFactory
+                .createSyncConfigurationBuilder(user, Constants.SYNC_SERVER_URL)
+                .build();
+        final SyncConfiguration adminConfig = configFactory
+                .createSyncConfigurationBuilder(adminUser, userConfig.getServerUrl().toString())
+                .build();
+
+        final CountDownLatch userInterruptLatch = new CountDownLatch(1);
+        final CountDownLatch adminInterruptLatch = new CountDownLatch(1);
+        Thread t = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm userRealm = Realm.getInstance(userConfig);
+                userRealm.beginTransaction();
+                userRealm.createObject(AllTypes.class);
+                userRealm.commitTransaction();
+                SyncSession userSession = SyncManager.getSession(userConfig);
+                try {
+                    // 1. Start download (which will be interrupted)
+                    userInterruptLatch.countDown();
```

Comment:
this will be a little bit flaky. there will still be a small chance the thread set to interrupted state after `downloadAllServerChanges()`. Actually the interrupted state can be set inside the current thread.
I think just call `Thread.currentThread().interrupt()` before the `downloadAllServerChanges()` would be enough and we can remove the latches.